File: 329-longest-increasing-path-in-a-matrix.java
class Solution {
    int max = 0;
    public int longestIncreasingPath(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }

        int[][] distance = new int[matrix.length][matrix[0].length];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                max = Math.max(max, dfs(matrix, distance, i, j));
            }
        }
        return max;
    }
    private int dfs(int[][] matrix, int[][] distance, int x, int y) {
        int[] directionX = {0, 0, 1, -1};
        int[] directionY = {1, -1, 0, 0};
        if (distance[x][y] != 0) {
            return distance[x][y];
        }
        for (int i = 0; i < 4; i++) {
            int nextX = x + directionX[i];
            int nextY = y + directionY[i];
            if (!isValid(matrix, nextX, nextY) || matrix[nextX][nextY] >= matrix[x][y]) {
                continue;
            }
            distance[x][y] = Math.max(distance[x][y], dfs(matrix, distance, nextX, nextY));
        }
        distance[x][y]++;
        return distance[x][y];
    }
    private boolean isValid(int[][] matrix, int x, int y) {
        return 0 <= x && x < matrix.length && 0 <= y && y < matrix[0].length;
    }
}

File: python3/329.longest-increasing-path-in-a-matrix.py
DIRECTIONS = (
    (0, 1),
    (0, -1),
    (1, 0),
    (-1, 0)
)

class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        n, m = len(matrix), len(matrix[0])
        memo = [[0] * m for _ in range(n)]
        
        def dfs(curr_x, curr_y):
            if memo[curr_x][curr_y] != 0:
                return memo[curr_x][curr_y]
            max_length = 0
            for diff_x, diff_y in DIRECTIONS:
                next_x = curr_x + diff_x
                next_y = curr_y + diff_y
                if in_bound(next_x, next_y, n, m) and matrix[next_x][next_y] > matrix[curr_x][curr_y]:
                    max_length = max(max_length, dfs(next_x, next_y))
            
            ans = max_length + 1
            memo[curr_x][curr_y] = ans
            return ans
        
        ans = 1
        
        for x in range(n):
            for y in range(m):
                ans = max(ans, dfs(x, y))
                
        return ans
    
def in_bound(x, y, n, m):
    return 0 <= x < n and 0 <= y < m
